### ISSUE TITLE
Feature Request : Bluetooth Proxy Support on BK7231x

### ISSUE THREAD

**Issue by rishabmehta7** [libretiny-eu/libretiny#162]
As Beken chips have bluetooth like ESP32, Having a BT Proxy would be amazing https://esphome.io/components/bluetooth_proxy.html

**Comment by jeison-souza** [comment]
This will help a lot, since tons of modules are using these chips right now !

**Comment by erapade** [comment]
Yes this would be amazing. For me it would mean that every lightbulb could be a potential Bluetooth proxy for less than $/€ 4. 

Anyone that knows of implementation is ongoing and how far away this is?

**Comment by rishabmehta7** [comment]
@kuba2k2 Any idea if anyone is working on this?

**Comment by kuba2k2** [comment]
Yes, nobody is working on this.

**Comment by rishabmehta7** [comment]
> Yes, nobody is working on this.

I have no idea what to do (I am ready to learn) else I can help in anyway I can.

**Comment by TuanAIoT** [comment]
Is there a way to get the bluetooth library file and compile it?
![6](https://github.com/libretiny-eu/libretiny/assets/127066596/b7e84fda-6059-4b6b-a88f-0ee01066ba14)


> Vâng, không ai đang làm việc này.

**Comment by tiimsvk** [comment]
no news in this thread?
Today I flashed one of these modules and to my surprise I found out that it also contains bluetooth

**Comment by erapade** [comment]
Did you managed to install a BT proxy as well. If so, I will probably pee I'm my pants :-)

**Comment by rishabmehta7** [comment]
> Did you managed to install a BT proxy as well. If so, I will most probably pee I'm my pants :-)

Nope. LibreTiny may not be a great platform. Bugs don't get addressed, no features get worked on. If OpenBeken finds a way to implement BTProxy. I am moving to that.

**Comment by kuba2k2** [comment]
See my reply in https://github.com/libretiny-eu/libretiny/issues/278#issuecomment-2213807829

**Comment by erapade** [comment]
> See my reply in https://github.com/libretiny-eu/libretiny/issues/278#issuecomment-2213807829

Does it mean that the code needs to be refactored before it's worth to add enough features to support BT proxy. 

As you probably have understood, what we are aiming for is to turn every light bulb into a BT proxy

**Comment by agittins** [comment]
> LibreTiny may not be a great platform. Bugs don't get addressed, no features get worked on. If OpenBeken finds a way to implement BTProxy. I am moving to that.

This is a pretty disrespectful attitude to spout in comments. Libretiny is not some corporate product that took your money and failed to deliver - it's a free and open source project that is run mostly by a single dev, and frankly it is amazing that such a thing exists!

I understand your frustration with the pace of change, but complaining about it like it's a failing is seriously uncool. It's not a failing, it's just the reality of the situation at hand. I was (and remain) super excited to discover an esphome-based implementation for beken chips, and I had spent some time in the openbeken side of things - I am so glad that Libretiny exists and has achieved as much as it has, and especially that it's easily incorporated into the standard esphome setup (for one thing, I don't have to have multiple places to store my iot codebases, awesome!).

Frankly, it's possible that the beken chips will never be able to do bluetooth proxy - the device might not have the grunt and memory to manage it (proxy is much more demanding than just sending broadcasts or supporting a single connection). I'd love to see bt proxy support happen (I have a vested interest, my Bermuda project) but the reality is it might not even be possible, so it's a significant investment for someone to dive into this to spend the time to possibly just confirm it can't be done - it's deeply unreasonable to criticise anyone for not choosing to take that gamble with their time.

People like kuba2k2 who put in serious work and offer it up to all of us for free deserve to be treated with respect, if you don't like their work pass on by, if you think it could be better, consider sponsoring them or someone else to help out, but eroding a dev's passion for something you directly benefit from by complaining about how much of their life they are donating to you is not helpful, and I'd encourage anyone to think compassionately about how they can best engage with the open source community in ways that benefit everyone and acknowledges contributions, rather than putting them down.

**Comment by rishabmehta7** [comment]
> > LibreTiny may not be a great platform. Bugs don't get addressed, no features get worked on. If OpenBeken finds a way to implement BTProxy. I am moving to that.
> 
> This is a pretty disrespectful attitude to spout in comments. Libretiny is not some corporate product that took your money and failed to deliver - it's a free and open source project that is run mostly by a single dev, and frankly it is amazing that such a thing exists!
> 
> I understand your frustration with the pace of change, but complaining about it like it's a failing is seriously uncool. It's not a failing, it's just the reality of the situation at hand. I was (and remain) super excited to discover an esphome-based implementation for beken chips, and I had spent some time in the openbeken side of things - I am so glad that Libretiny exists and has achieved as much as it has, and especially that it's easily incorporated into the standard esphome setup (for one thing, I don't have to have multiple places to store my iot codebases, awesome!).
> 
> Frankly, it's possible that the beken chips will never be able to do bluetooth proxy - the device might not have the grunt and memory to manage it (proxy is much more demanding than just sending broadcasts or supporting a single connection). I'd love to see bt proxy support happen (I have a vested interest, my Bermuda project) but the reality is it might not even be possible, so it's a significant investment for someone to dive into this to spend the time to possibly just confirm it can't be done - it's deeply unreasonable to criticise anyone for not choosing to take that gamble with their time.
> 
> People like kuba2k2 who put in serious work and offer it up to all of us for free deserve to be treated with respect, if you don't like their work pass on by, if you think it could be better, consider sponsoring them or someone else to help out, but eroding a dev's passion for something you directly benefit from by complaining about how much of their life they are donating to you is not helpful, and I'd encourage anyone to think compassionately about how they can best engage with the open source community in ways that benefit everyone and acknowledges contributions, rather than putting them down.

I'd agree I was maybe a tad bit over the line, but that's my passion and love for the project.
I do contribute in any way I can.
What I meant by bugs don't get fixed - here's an example https://github.com/libretiny-eu/libretiny/issues/280.

The frustration is not towards Kuba2k2 but more about even tho LT is added to ESPhome it's not getting the attention and love the main line gets. We did discuss a bit about it on the other bug linked above.

**Comment by agittins** [comment]
I agree with you that it would be nice to see a way for the esphome team to chip in here - it seems like a lot of the heavy lifting to get bk chips as a platform in esphome is done and benefits esphome already. I noticed your comment on the other bug and that you have reached out to esphome to see how they can contribute, which is awesome.

I know I came on a bit strong, in part I'm reacting to a pattern of OSS devs getting burnt-out and it's clear that you care about the project.

**Comment by kuba2k2** [comment]
I'm not sure I'd say I'm burnt-out with this project. I definitely will work on LibreTiny much more in the future, just don't know when. Right now I'm working on another big project of mine, which I've had for nearly 6 years now, and which needs attention after more than a year of no maintenance whatsoever. When I finish that, I see no reason not to get back to LibreTiny (unless a different particular "commercial" project comes up, I can't tell yet).

**Comment by erapade** [comment]
> Did you managed to install a BT proxy as well. If so, I will most probably pee I'm my pants :-)

So my pants will stay dry at least for some months/years... 😅

**Comment by Cossid** [comment]
Bluetooth at its core is apparently working in LibreTiny on Beken (at a base proof-of-concept level).  The heavy lift here is in ESPHome, as the current implementation is very ESP32 dependent.  ESPHome's ble_client and bluetooth_proxy would need to be rewritten into agnostic standardized contracts/headers, and then it would be easier to implement platform-specific versions of it.  Prior to LibreTiny, ESP32 was the only platform which had bluetooth.  Whoever takes on such a project for any new platform would need to include that refactor, unless it is addressed by ESPHome in advance.  So far no one has volunteered, and there hasn't been enough user input to increase priority.

**Comment by rishabmehta7** [comment]
> Bluetooth at its core is apparently working in LibreTiny on Beken (at a base proof-of-concept level).  The heavy lift here is in ESPHome, as the current implementation is very ESP32 dependent.  ESPHome's ble_client and bluetooth_proxy would need to be rewritten into agnostic standardized contracts/headers, and then it would be easier to implement platform-specific versions of it.  Prior to LibreTiny, ESP32 was the only platform which had bluetooth.  Whoever takes on such a project for any new platform would need to include that refactor, unless it is addressed by ESPHome in advance.  So far no one has volunteered, and there hasn't been enough user input to increase priority.

In that case if you know what has to be said and where, we can start putting pressure so things can move in the right direction...

**Comment by Bl00d-B0b** [comment]
Big progress on this — now ready for review on the ESPHome side:

- esphome/esphome#17150 — platform-neutral BLE layer: all advertisement-based BLE sensor platforms (BTHome, xiaomi, ruuvi, ble_rssi, presence, …) work on any chip with a tracker
- esphome/esphome#17135 — `bk72xx_ble_tracker`: BLE 5.x scanner for BK7231N / BK7238 on the bundled LibreTiny toolchain
- esphome/esphome#16691 — same for LN882H

Hardware-validated in production on BK7231N (CB3S) and LN882H (CB3S) devices, including encrypted BTHome. The Bluetooth *proxy* itself is the follow-up step on the same contract — the trackers already expose the raw-advertisement callback it needs.

If you want this, a 👍 on the ESPHome PRs helps the maintainers prioritize.

**Comment by fightforlife** [comment]
It seems all preconditions are merged, is there a ticket for the proxy support that we can follow?

Found: https://github.com/orgs/esphome/discussions/3758

**Comment by rishabmehta7** [comment]
It was all implemented in the latest release.

**Comment by fightforlife** [comment]
Bad news: I flashed the latest firmware with Bluetooth Proxy to a device (cb2s) that ran fine for years with multiple firmware iterations.
This morning it is broken, no reaction wat so ever, even on hard reboot.
I hope it is just random, but I will leave this note here if someone else has this problem.

**Comment by Dr-Blank** [comment]
> This morning it is broken

device bricking on bk7283 is a known issue when enabling ble support. see https://github.com/libretiny-eu/libretiny/issues/407

but yours is cb2s which is BK7231N and ble is working on many of me cb3s devices which have the same chip. yours is not booting or crashing aftr booting?

**Comment by kuba2k2** [comment]
@fightforlife The best thing you can do now is to pull it apart, check UART logs and dump the flash memory as it is. Otherwise we won't know what caused the issue. If the device doesn't boot you'll have to use UART anyway.

**Comment by Bl00d-B0b** [comment]
@fightforlife The overnight timing makes this look like the [#407](https://github.com/libretiny-eu/libretiny/issues/407) mechanism on BK7231N's schedule. On BK7238 the BDK bond store erases at BLE bring-up, which is why those bricks are immediate. Hardware testing on the fix branch ([#408](https://github.com/libretiny-eu/libretiny/pull/408), report by @Dr-Blank there) showed BK7231N *defers* the bond-store write instead — so a scan-only proxy runs fine until something first writes the store, and with stock v1.13.0 the unconfigured store address resolves to sector 0: bootloader erased hours after a perfectly healthy boot, dead on the next reset.

One-look confirmation from your flash dump: read `0x0-0x1000`. If the #407 mechanism hit, the bootloader sector is mostly `0xFF` (on BK7238 it was 3980 of 4096 bytes erased); an intact dump rules this out and the search moves elsewhere.

If it matches: recovery is a UART reflash of the full image (the rest of the flash is typically untouched), and [#408](https://github.com/libretiny-eu/libretiny/pull/408) is the fix — it gives every layout a dedicated `ble_bonding` partition and fails the build rather than allowing a zero-length store. It is hardware-validated on BK7238 (T1-3S, the immediate-brick case) and BK7231N (CB3S: full-flash diff clean, bond sector provably untouched). Until it merges and ships, BLE-enabled BK builds can pin the fixed platform:

```yaml
platformio_options:
  platform: https://github.com/Bl00d-B0b/libretiny.git#fix/bk72xx-ble-bonding-partition
```

With bk72xx BLE now in an ESPHome release while LibreTiny v1.13.0 still carries the unfixed store, more of these reports are likely — worth prioritizing #408 for the next LibreTiny release.

**Comment by fightforlife** [comment]
I did not have the time yet to solder the UART-USB Adapter.
But from the HA Logs I could see that the device went offline 6min 50s after the successful flash. It is a BK7231N.
And by powering it directly with 3.3V, it was able to boot. So it seems the bootloader is working, and it might be a different bug. (could be hardware)

**Comment by Bl00d-B0b** [comment]
Production stability datapoint: 21 BK7231x devices (BK7231N-based thermostats, light switches, LED strip controllers and power sockets, incl. T34 modules) run `bluetooth_proxy` here, all through one shared include on stock ESPHome (the BK72xx tracker ships in-tree since 2026.8). Scanning is gated on the Home Assistant API connection - the tracker boots idle and only scans while HA is actually listening, which also keeps the shared WiFi/BLE radio free otherwise:

```yaml
bluetooth_proxy:

bk72xx_ble_tracker:
  scan_parameters:
    continuous: false    # boot idle; scan only while HA is connected

api:
  on_client_connected:
    - bk72xx_ble_tracker.start_scan:
        continuous: true
  on_client_disconnected:
    # fires for EVERY departing client (log viewer, CLI) - only stop
    # scanning when the LAST client is gone
    - if:
        condition:
          not:
            api.connected:
        then:
          - bk72xx_ble_tracker.stop_scan:
```

Current numbers: all 21 show 7.6 days uptime, i.e. untouched since the last planned fleet OTA on 2026-08-31, scanning continuously the whole time (HA stays connected) with zero unplanned reboots or BLE errors. The proxy has been on this fleet since early August. Default scan timings throughout - no per-device tuning needed.

**Comment by fightforlife** [comment]
It seems my issue is something else. My board is booting fine after disconnecting it from power for a few seconds.
But I can reproduce the crash after 6-7 minuted reliably. Will try to get verbose logs.

This is the last log before it goes offline.
```
[20:00:30.464][VV][api.service:013]: send_message bluetooth_le_raw_advertisements_response: BluetoothLERawAdvertisementsResponse {
[20:00:30.464][VV][api.service:013]:     advertisements: BluetoothLERawAdvertisement {
[20:00:30.464][VV][api.service:013]:   address: 122668798630280
[20:00:30.464][VV][api.service:013]:   rssi: -62
[20:00:30.464][VV][api.service:013]:   address_type: 1
[20:00:30.464][VV][api.service:013]:   data: 03:03:9F:FE:17:16:9F:FE:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
[20:00:30.464][VV][api.service:013]:  }
[20:00:30.464][VV][api.service:013]:  }
[20:00:30.465][VV][bluetooth_proxy:156]: Sent batch of 1 BLE advertisements
[20:00:30.477][D][wifi:2485]: Roam scan (-49 dBm, attempt 1/3)
```

Discussed here: https://github.com/esphome/esphome/issues/18592
